runtests.py

```diff
@@ -1002,10 +1002,18 @@ def source_files(self, workdir, module_name, file_list):
 
     def split_source_and_output(self, test_directory, module, workdir):
         source_file = self.find_module_source_file(os.path.join(test_directory, module) + '.pyx')
-        with io_open(source_file, 'r', encoding='ISO-8859-1') as source_and_output:
+
+        from Cython.Utils import detect_opened_file_encoding
+        with io_open(source_file, 'rb') as f:
+            encoding = detect_opened_file_encoding(f, default='ISO-8859-1')
+        if encoding.lower() == 'ascii':
+            encoding = 'ISO-8859-1' # at least one test is based around a file tagged with ascii
+            # but with a character that can't be read with ascii. Therefore use different default
```

Comment:
I think this is a very good test, but it's also a bit risky to play with the encoding here. This code was written to provide a pass-through, and anything we change here might make tests (even existing tests) silently test something other than they were written for. In some error test cases, the encoding declaration is also intentionally wrong, IIRC.

Instead, maybe we could decode only the error messages (see `ErrorWriter`), and not the entire test source file?

Reply:
How about if I add an alternate way of specifying the errors so it first searches for `whatever_the_filename_is.errors`? I imagine most of the time it's still better to keep the errors with the code, but that seems more robust than trying to guess the source file encoding (which, as you say, can be intentionally wrong).